public/app/features/logs/components/PopoverMenu.tsx

```diff
@@ -0,0 +1,88 @@
+import { css } from '@emotion/css';
+import React, { useEffect, useRef } from 'react';
+
+import { GrafanaTheme2, LogRowModel } from '@grafana/data';
+import { Menu, useStyles2 } from '@grafana/ui';
+
+import { copyText } from '../utils';
+
+interface PopoverMenuProps {
+  selection: string;
+  x: number;
+  y: number;
+  onClickFilterValue?: (value: string, refId?: string) => void;
+  onClickFilterOutValue?: (value: string, refId?: string) => void;
+  row: LogRowModel;
+  close: () => void;
+}
+
+export const PopoverMenu = ({
+  x,
+  y,
+  onClickFilterValue,
+  onClickFilterOutValue,
+  selection,
+  row,
+  close,
+}: PopoverMenuProps) => {
+  const containerRef = useRef<HTMLDivElement | null>(null);
+  const styles = useStyles2(getStyles);
+
+  useEffect(() => {
+    function handleEscape(e: KeyboardEvent) {
+      if (e.key === 'Escape') {
+        close();
+      }
+    }
+    document.addEventListener('keyup', handleEscape);
+
+    return () => {
+      document.removeEventListener('keyup', handleEscape);
+    };
+  }, [close]);
+
+  const supported = onClickFilterValue || onClickFilterOutValue;
+
+  if (!supported) {
+    return null;
+  }
+
+  return (
+    <div className={styles.menu} style={{ top: y, left: x }}>
+      <Menu ref={containerRef}>
+        <Menu.Item
+          label="Copy selection"
+          onClick={() => {
+            copyText(selection, containerRef);
+            close();
+          }}
+        />
+        {onClickFilterValue && (
```

Comment:
I think we may need a better API to check what types of filters the given data source supports. At the moment we have a single check in https://github.com/grafana/grafana/blob/main/public/app/features/explore/Logs/LogsContainer.tsx#L91 but a datasource may support only label/tag filter (`ADD_FILTER / ADD_FILTER_OUT`) but not line filters (`ADD_LINE_FILTER / ADD_LINE_FILTER_OUT`) which may lead to showing no-op items in the menu.

We did a very similar thing with Supplementary Queries. A data source implements `getSupportedSupplementaryQueryTypes()` providing list of supported types.

Reply:
I think at some point we should deprecate checking for `modifyQuery` in favor of `hasToggleableQueryFiltersSupport()` which is the desired behavior (and API) for log details filters.

Reply:
Are you making more changing around this in this PR? I think at the moment the popover will show for data sources that may not support `ADD_LINE_FILTER` but they support `ADD_FILTER` 🤔

Reply:
I will follow up this PR with a fix for this. Something similar to `getSupportedSupplementaryQueryTypes()` would make an interesting approach, as I don't think we need a new interface for this.

Reply:
Will be addressed by https://github.com/grafana/grafana/pull/78322